### ISSUE TITLE
Is tf2xla  used even when jit_compile=False

### ISSUE THREAD

**Issue by whatdhack** [tensorflow/tensorflow#55207]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):Ubuntu
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.8
- Python version: 3.9.10
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
With TF_CPP_MAX_VLOG_LEVEL=3 ,  still see lines like the following  in the log.

> 10048:2022-03-11 09:50:05.273639: I tensorflow/compiler/tf2xla/xla_op_registry.cc:328] XLA op registration: device: XLA_GPU_JIT op: XlaWhile
10049:2022-03-11 09:50:05.273651: I tensorflow/compiler/tf2xla/xla_op_registry.cc:328] XLA op registration: device: XLA_CPU_JIT op: XlaWhile
10050:2022-03-11 09:50:05.273660: I tensorflow/compiler/tf2xla/xla_op_registry.cc:328] XLA op registration: device: XLA_GPU_JIT op: XlaIf
10051:2022-03-11 09:50:05.273670: I tensorflow/compiler/tf2xla/xla_op_registry.cc:328] XLA op registration: device: XLA_CPU_JIT op: XlaIf
10052:2022-03-11 09:50:05.273680: I tensorflow/compiler/tf2xla/xla_op_registry.cc:328] XLA op registration: device: XLA_GPU_JIT op: StatelessIf
10053:2022-03-11 09:50:05.273689: I tensorflow/compiler/tf2xla/xla_op_registry.cc:328] XLA op registration: device: XLA_CPU_JIT op: StatelessIf
10054:2022-03-11 09:50:05.273698: I tensorflow/compiler/tf2xla/xla_op_registry.cc:328] XLA op registration: device: XLA_GPU_JIT op: If
10055:2022-03-11 09:50:05.273707: I tensorflow/compiler/tf2xla/xla_op_registry.cc:328] XLA op registration: device: XLA_CPU_JIT op: If
10056:2022-03-11 09:50:05.273718: I tensorflow/compiler/tf2xla/xla_op_registry.cc:328] XLA op registration: device: XLA_GPU_JIT op: StatelessCase
10057:2022-03-11 09:50:05.273727: I tensorflow/compiler/tf2xla/xla_op_registry.cc:328] XLA op registration: device: XLA_CPU_JIT op: StatelessCase
.
.
.
10973:2022-03-11 09:50:05.281066: I tensorflow/compiler/tf2xla/xla_op_registry.cc:414] For operation Reshape required constants are: shape
10974:2022-03-11 09:50:05.281070: I tensorflow/compiler/tf2xla/const_analysis.cc:261] marking consts for must-be-const inputs of model/flatten/Reshape
10975:2022-03-11 09:50:05.281074: I tensorflow/compiler/tf2xla/const_analysis.cc:222] marking consts for must-be-const node model/flatten/Const
10978:2022-03-11 09:50:05.281228: I tensorflow/compiler/tf2xla/xla_op_registry.cc:153] tf_xla_cpu_global_jit = 0
10979:2022-03-11 09:50:05.281237: I tensorflow/compiler/tf2xla/xla_op_registry.cc:51] LaunchOpHasKernelForDevice kernel_class_name: XlaLocalLaunchOp
10980:2022-03-11 09:50:05.281247: I tensorflow/compiler/tf2xla/xla_op_registry.cc:51] LaunchOpHasKernelForDevice kernel_class_name: XlaLocalLaunchOp
11170:2022-03-11 09:50:05.282582: I tensorflow/compiler/tf2xla/xla_op_registry.cc:414] For operation Sum required constants are: reduction_indices
11171:2022-03-11 09:50:05.282588: I tensorflow/compiler/tf2xla/const_analysis.cc:261] marking consts for must-be-const inputs of sparse_categorical_crossentropy/weighted_loss/Sum
11172:2022-03-11 09:50:05.282605: I tensorflow/compiler/tf2xla/xla_op_registry.cc:414] For operation Reshape required constants are: shape
11173:2022-03-11 09:50:05.282611: I tensorflow/compiler/tf2xla/const_analysis.cc:261] marking consts for must-be-const inputs of model/flatten/Reshape

**Describe the expected behavior**

tf2xla should not show up in the log.

**Comment by chunduriv** [comment]
@whatdhack, 

In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here.

Meanwhile, can you refer this [comment](https://github.com/tensorflow/tensorflow/issues/54925#issuecomment-1059028169) and let us know if it helps? Thanks!

**Comment by whatdhack** [comment]
Hi, I think it is coming from the following.  Looks like this is not guarded by any check for whether XLA is asked for or not. 
  
https://github.com/tensorflow/tensorflow/blob/3f878cff5b698b82eea85db2b60d65a2e320850e/tensorflow/compiler/jit/jit_compilation_pass_registration.cc#L35

What I am trying to answer if this pollutes non-XLA executions with XLA  kernels.

**Comment by gadagashwini** [comment]
@whatdhack,
On colab i dont see any XLA related logs if I make `jit_compile=False `. 
Take a look at [gist](https://colab.research.google.com/drive/1OpxbEb9csWAIlzNCFX8HG-xCt7CqFu8e?usp=sharing). Thanks!

**Comment by whatdhack** [comment]
@gadagashwini , thanks for following up.  I made a slight change to your code as follows. 

```
import tensorflow as tf

@tf.function(jit_compile=False)
def func(a, b):
    for i in tf.range(10):
        c = a + b
        a = b
        b = c 
    return c
concrete_fn = func.get_concrete_function(
    tf.TensorSpec(shape=[3], dtype=tf.float32, name='a'),
    tf.TensorSpec(shape=[3], dtype=tf.float32, name='b')
    )
a = tf.random.normal([3])
b = tf.random.normal([3])

c = concrete_fn(a,b)

```
Run the above code as follows and you will see many calls to tf2xla and xla.

TF_CPP_MAX_VLOG_LEVEL=3 python ./xla_gist.py

**Comment by gadagashwini** [comment]
@whatdhack, 
I tried with given code on colab, i didn’t see XLA logs. 
Please take a look at [gist](https://colab.research.google.com/drive/1go4KJvraudH35S1EWH0GmNsjIuMqhLYR?usp=sharing). Thanks!

**Comment by whatdhack** [comment]
Did you set TF_CPP_MAX_VLOG_LEVEL=3 ?

**Comment by gadagashwini** [comment]
@whatdhack,
I tried on Colab. Added the code with log 
```
TF_CPP_MAX_VLOG_LEVEL=3
import tensorflow as tf


@tf.function(jit_compile=False)
def func(a, b):
    for i in tf.range(10):
        c = a + b
        a = b
        b = c 
    return c
concrete_fn = func.get_concrete_function(
    tf.TensorSpec(shape=[3], dtype=tf.float32, name='a'),
    tf.TensorSpec(shape=[3], dtype=tf.float32, name='b')
    )
a = tf.random.normal([3])
b = tf.random.normal([3])

c = concrete_fn(a,b)
c
```
`<tf.Tensor: shape=(3,), dtype=float32, numpy=array([-189.39798 ,   98.73976 ,   34.542355], dtype=float32)>`

**Comment by whatdhack** [comment]
@gadagashwini , thanks. I do not see the log, only see the output. Looks like there are issues with seeing logs in colab per - https://github.com/tensorflow/tensorflow/issues/31870  .  You will need to use a cloud or on-premise instance to see the log.

**Comment by sachinprasadhs** [comment]
When the `jit_compile` is set to False XLA won't be used to perform operations.
I tried to reproduce the error using your code in colab by running the code as python script. 
I was able to see some logs, but it did not have any trace of usage of XLA. 
[Here](https://colab.sandbox.google.com/gist/sachinprasadhs/ad87e5a5065d3a46afc78ebde6f39d91/55207.ipynb) is the attached gist for reference. Thanks!

**Comment by whatdhack** [comment]
Thanks for enabling the log in colab.  There following are the 4 reference to xla I  could see in the log, 

```
2022-04-18 19:39:22.726818: W tensorflow/core/util/dump_graph.cc:134] Failed to dump encapsulate_xla_computations_before because dump location is not  specified through either TF_DUMP_GRAPH_PREFIX environment variable or function argument.
2022-04-18 19:39:22.727244: W tensorflow/core/util/dump_graph.cc:134] Failed to dump encapsulate_xla_computations_halfway because dump location is not  specified through either TF_DUMP_GRAPH_PREFIX environment variable or function argument.
2022-04-18 19:39:22.727439: W tensorflow/core/util/dump_graph.cc:134] Failed to dump encapsulate_xla_computations_after because dump location is not  specified through either TF_DUMP_GRAPH_PREFIX environment variable or function argument.


2022-04-18 19:39:22.753706: W tensorflow/core/util/dump_graph.cc:134] Failed to dump build_xla_ops because dump location is not  specified through either TF_DUMP_GRAPH_PREFIX environment variable or function argument.
```

**Comment by sachinprasadhs** [comment]
It is mentioned as failed to dump `encapsulate_xla_computations` for before, halfway and after in the warning log, it means it is not making use of XLA. 
You can ignore these warnings.

**Comment by whatdhack** [comment]
That dump failed because the TF_DUMP_GRAPH_PREFIX was not populated ?  I do not  think these logs can be ignored. My suspicion is under the hood XLA is enabled, even though it was not supposed to be.

**Comment by david-thrower** [comment]
I am having the same issue with transfer learning from models from keras_nlp and from tensorflow hub.

Example:

```python3
preprocessor = hub.KerasLayer(
    "https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3")
encoder_inputs = preprocessor(text_input)
encoder = hub.KerasLayer(
    "https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4",
    trainable=True)
outputs = encoder(encoder_inputs)
pooled_output = outputs["pooled_output"]      # [batch_size, 768].
sequence_output = outputs["sequence_output"]  # [batch_size, seq_length, 768].
embedding_model = tf.keras.Model(text_input, pooled_output)
```
I use the above as a foundation model for a binary classification model `foo` ...:

`foo.fit(...)`

Gives the error seen below (despite jit_compile=False) being explicitly passed to both `foo.compile()` and to the optimizer.

```
2023-11-17 21:55:54.272679: E tensorflow/compiler/xla/status_macros.cc:54] INTERNAL: RET_CHECK failure (tensorflow/compiler/jit/xla_launch_util.cc:563) variable.var() && variable.var()->is_initialized 
```
The issue appears in  v2.14.0 with AdamW and Lion optimizers. In 2.12.0 with Adam optimizer, the issue is absent.

**Comment by david-thrower** [comment]
Cancel on the last comment. Digging deeper, I found where the param jit_compile was not being forwarded, and once corrected, it appears the issue was fixed. .. But there is another Issue that (my efforts to resolve) led me here ...  I will look for this issue and and attach a link to it for reference. The issue I am still seeing is It an issue relating to where for some foundation models, if you set some layers as trainable and others not trainable, it will throw an error, that I believe other commenters have found affects BatchNormalization layers. Sorry for the oversight.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
